File: day4/aoc.nim
import std/sets
from std/strutils import isDigit,parseInt
from std/sequtils import toSeq

proc p1=
    let input = open("input4.txt")
    defer: input.close()
    var result = 0
    for line in lines(input):
        var 
            str = ""
            numbers:array[4,int]

        #We read all numbers in the line
        var i = 0
        for l in line:
            if isDigit(l):
                str.add(l)
            else:
                numbers[i] = parseInt(str)
                str = ""
                i += 1
        numbers[i] = parseInt(str)

        #Now we get the ranges and verify if at least one is a a subset of the other
        let 
            s1 = toHashSet(
            toSeq(numbers[0]..numbers[1])
            )
            s2 = toHashSet(
            toSeq(numbers[2]..numbers[3])
            )

        if s1 <= s2 or s2 <= s1:
            result += 1
    echo $result, " pairs members fully contains the other"

proc p2=
    let input = open("input4.txt")
    defer: input.close()
    var result = 0
    for line in lines(input):
        var 
            str = ""
            numbers:array[4,int]

        #We read all numbers in the line
        var i = 0
        for l in line:
            if isDigit(l):
                str.add(l)
            else:
                numbers[i] = parseInt(str)
                str = ""
                i += 1
        numbers[i] = parseInt(str)

        #Now we get the ranges and verify if at least one is a a subset of the other
        let 
            s1 = toHashSet(
            toSeq(numbers[0]..numbers[1])
            )
            s2 = toHashSet(
            toSeq(numbers[2]..numbers[3])
            )

        if len(intersection(s1, s2)) != 0:
            result += 1
    echo $result, " pairs! intersect"

when isMainModule:
    p1()
    p2()

File: day4/aoc.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define min(X, Y) (((X) < (Y)) ? (X) : (Y))
#define max(X, Y) (((X) > (Y)) ? (X) : (Y))

char** getTokens(char of[], const char token[], int nbToken) {
    char** result = malloc(sizeof(char*) * nbToken);

    result[0] = strtok(of, token);
    for (int i = 1; i < nbToken; i++) {
        result[i] = strtok(NULL, token);
    }
    return result;
}

typedef struct _Set {
    int from;
    int to;
}Set;

Set newSet(int from, int to) {
    Set result;
    result.from = from;
    result.to = to;
    return result;
}

int size(const Set a) {
    return a.to - a.from + 1;
}

Set intersection(const Set a, const Set b) {
    Set result;
    result.from = max(a.from, b.from);
    result.to = min(a.to, b.to);
    return result;
}

Set uunion(const Set a, const Set b) {
    Set result;
    result.from = min(a.from, b.from);
    result.to = max(a.to, b.to);
    return result;
}


void p1(FILE* fptr) {
    const size_t line_size = 300 * sizeof(char); //Max Size of a line
    char* line = malloc(line_size * sizeof(char));
    int nbFullyContains = 0;
    while (fgets(line, line_size, fptr)) {
        int sets[4];

        char** tokens = getTokens(line, ",", 2);
        int count = 0;
        for (int i = 0; i < 2; i++) {
            char** subTok = getTokens(tokens[i], "-", 2);
            for (int j = 0; j < 2; j++) {
                sscanf(subTok[j], "%d", &(sets[count]));
                count++;
            }
        }
        Set a = newSet(sets[0], sets[1]);
        Set b = newSet(sets[2], sets[3]);

        if (size(intersection(a,b)) == min(size(a), size(b))) {
            nbFullyContains++;
        }
        free(tokens);
    }
    printf("%d pairs fully contains each other\n", nbFullyContains);
    free(line);
}

void p2(FILE* fptr) {
    const size_t line_size = 300 * sizeof(char); //Max Size of a line
    char* line = malloc(line_size * sizeof(char));
    int nbOverlaps = 0;
    while (fgets(line, line_size, fptr)) {
        int sets[4];

        char** tokens = getTokens(line, ",", 2);
        int count = 0;
        for (int i = 0; i < 2; i++) {
            char** subTok = getTokens(tokens[i], "-", 2);
            for (int j = 0; j < 2; j++) {
                sscanf(subTok[j], "%d", &(sets[count]));
                count++;
            }
        }
        Set a = newSet(sets[0], sets[1]);
        Set b = newSet(sets[2], sets[3]);

        if (size(intersection(a,b)) > 0) {
            nbOverlaps++;
        }
        free(tokens);
    }
    printf("%d pairs overlaps\n", nbOverlaps);
    free(line);
}

int main() {
    FILE* fptr;
    
    fptr = fopen("input4.txt", "r");
    if (!fptr) {
        printf("Failed to open file");
        return -1;
    }
    p1(fptr);
    fclose(fptr);

    fptr = fopen("input4.txt", "r");
    if (!fptr) {
        printf("Failed to open file");
        return -1;
    }
    p2(fptr);
    fclose(fptr);

    return 0;
}